File: solutions/javascript/meetup/1/meetup.js

const DAYS_OF_WEEK = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];

// Computes the number of days in a month.
function daysInMonth(year, month) {
  
  // Months in Javascript start at 0.
  
  // february (28 or 29 depending of the year)
  if (month == 1) {
    // check if the year is bisextile
    return ((year %4 == 0) && ((year % 400 != 0) || (year % 100 ==0))) ? 29 : 28;
  }

  // 31 days month (January, March, May, July, August, October, December)
  return ([0, 2, 4, 6, 7, 9, 11].includes(month)) ? 31 : 30;
}

// Computes how many days forwards the dayOfWeek ('Sunday' to 'Saturday') lays from the numeric dayOfMonth.
function offset(year, month, dayOfMonth, dayOfWeek) {
  
  const date = new Date(year, month, dayOfMonth);
  let offset = DAYS_OF_WEEK.indexOf(dayOfWeek) - date.getDay();
  if (offset < 0) offset += 7;
  return offset;
}

// Returns the Date of a meetup based on the type ('first', 'second', 'third', 'fourth', 'last', and 'teenth')
// and the desired DayOfWeek.
export function meetup(year, month, type, dayOfWeek) {
  
  // adjust month to start at 0 (JS convention)
  month -= 1;

  switch (type) {
    case 'teenth':
      // find which dayOfWeek is the 13th of this month and adjust forward to match the specified dayOfWeek.
      return new Date(year, month, 13+offset(year, month, 13, dayOfWeek));
    case 'first':
      // find which day is the first day of the month and adjust forward to match the specified day parameter.
      return new Date(year, month, 1+offset(year, month, 1, dayOfWeek));
    case 'second':
      // find which day is the first day of the month and adjust forward to match the specified day parameter,
      // then skip by 7 days to get the second day of the month.
      return new Date(year, month, 8+offset(year, month, 1, dayOfWeek));
    case 'third':
      // find which day is the first day of the month and adjust forward to match the specified day parameter,
      // then skip by 14 days to get the third day of the month .
      return new Date(year, month, 15+offset(year, month, 1, dayOfWeek));
    case 'fourth':
      // find which day is the first day of the month and adjust forward to match the specified day parameter,
      // then skip by 21 days to get the third day of the month .
      return new Date(year, month, 22+offset(year, month, 1, dayOfWeek));
    case 'last':
      // find which day is the last day of the month, adjust backward to match the specified day parameter.
      let lastDay = daysInMonth(year, month);
      let lastOffset = offset(year, month, lastDay, dayOfWeek);
      // unless the offset is 0, we are now in the next month, back out a week
      if (lastOffset > 0) lastOffset -= 7;
      return new Date(year, month, lastDay+lastOffset);
    default:
      return "";
  }
};

File: solutions/go/meetup/1/meetup.go
package meetup

import "time"

// WeekSchedule defines the type used to request a specific weekday of the month
// as first, second, third, fourth, teenth (between the 13th and the 19th),
// and last (different from fourth when there is 5 of the same weekday in a month).
type WeekSchedule int

const (
    First WeekSchedule = iota
    Second
    Third
    Fourth
    Teenth
    Last
)

// DaysInMonth represents the number of days in a month. It doesn't account for
// leap years.
var daysInMonth = map[time.Month]int{
    time.January: 31, time.February: 28, time.March: 31, time.April: 30, time.May: 31,
    time.June: 30, time.July: 31, time.August: 31, time.September: 30,
    time.October: 31, time.November: 30, time.December: 31}

// IsLeapYear checks if a specific year is a leap year.
func isLeapYear(year int) bool {
    return (year % 4 == 0) && (!(year % 100 == 0) || (year % 400 == 0))
}

// LastDayForMonth computes the number of the last day in a specific month.
// It accounts for leap years.
func lastDayForMonth(month time.Month, year int) int {
    res := daysInMonth[month]
    if month == time.February && isLeapYear(year) {
        res += 1
    }
    return res
}

// Day returns the day of the month where a specific weekday occurs. The specific
// weekday is requested using a WeekSchedule argument (see above for definition of
// WeekSchedule).
func Day(wSched WeekSchedule, wDay time.Weekday, month time.Month, year int) int {
    firstDay := time.Date(year, month, 1, 0, 0, 0, 0, time.UTC)
    // Start with the first requested weekday in the month.
    dayOfMonth := 1 + int(wDay) - int(firstDay.Weekday())
    if dayOfMonth <= 0 {
        dayOfMonth += 7
    }
    // If the request is not for the first weekday, adjust accordingly.
    switch wSched {
        case Second: dayOfMonth += 7
        case Third: dayOfMonth += 14
        case Fourth: dayOfMonth += 21
        case Last:
        	dayOfMonth += 21
        	if lastDayForMonth(month, year) - dayOfMonth >= 7 {
                dayOfMonth += 7
            }
        case Teenth:
        	for dayOfMonth < 13 {
                dayOfMonth += 7
            }
    }
	return dayOfMonth
}
